File: hooks/tool-guard.py
#!/usr/bin/env python3
"""preToolUse hook: protect the workflow-state file from direct mutations.

Fires before any tool executes. The hook keeps `permission=allow` for almost
every call. It only sets `permission=ask` when a non-shell editing tool
(Write, Edit, MultiEdit, NotebookEdit) targets a file whose basename is
`workflow-state.json`, so the user can confirm before bypassing the
workflow-state CLI or MCP writer helpers with raw JSON edits.

The hook never denies a request and never modifies tool input. Shell commands
are out of scope here; `hooks/shell-guard.py` covers `beforeShellExecution`.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path, PurePosixPath

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _trace import trace as _trace  # noqa: E402
from _active_role import (  # noqa: E402
    agent_is_readonly,
    agent_tools_allowlist,
    get_active_role,
)
from _tool_payload import extract_file_path, extract_tool_name  # noqa: E402


EDIT_TOOLS = {"Write", "Edit", "MultiEdit", "NotebookEdit"}


def _read_payload() -> dict:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return {"_invalid_json": True}
    return value if isinstance(value, dict) else {"payload": value}


def main() -> int:
    payload = _read_payload()
    if payload.get("_invalid_json"):
        print(json.dumps({"status": "pass", "fail_open": True, "permission": "allow", "user_message": "Tool-guard input was not JSON; skipped."}))
        return 0

    tool_name = extract_tool_name(payload)
    file_path = extract_file_path(payload)
    basename = PurePosixPath(file_path.replace("\\", "/")).name if file_path else ""

    active_role = get_active_role()
    role_readonly = agent_is_readonly(active_role) if active_role else None
    role_tools = agent_tools_allowlist(active_role) if active_role else None

    if tool_name in EDIT_TOOLS and basename == "workflow-state.json":
        permission = "ask"
        message = (
            "Tool-guard observed a direct edit to a workflow-state.json document. "
            "Prefer the cursor-state-bridge tools or scripts/workflow-state.py "
            "(the installed `.cursor/state/workflow-state.py` path is a compatibility shim) so phase, "
            "acceptance criteria, and history advance through one bounded path."
        )
        status = "ask"
    elif active_role and role_readonly is True and tool_name in EDIT_TOOLS:
        permission = "ask"
        message = (
            f"Tool-guard: active subagent role `{active_role}` declares `readonly: true` "
            f"in agents/{active_role}.md. Edit-class tools require user confirmation. "
            "If this edit is intended, approve to proceed; otherwise route the change to a non-readonly role."
        )
        status = "ask"
    elif active_role and isinstance(role_tools, list) and tool_name and tool_name not in role_tools:
        permission = "ask"
        message = (
            f"Tool-guard: active subagent role `{active_role}` does not list `{tool_name}` in its "
            f"`tools:` allowlist (agents/{active_role}.md). Approve to proceed or extend the allowlist."
        )
        status = "ask"
    else:
        permission = "allow"
        message = "Tool-guard saw no protected file mutation."
        status = "pass"

    output = {
        "status": status,
        "fail_open": permission != "deny",
        "permission": permission,
        "user_message": message,
        "tool_name": tool_name,
        "file_path": file_path,
    }
    _trace({
        "hook": "tool-guard",
        "event": "preToolUse",
        "status": status,
        "permission": permission,
        "tool_name": tool_name,
        "file_basename": basename,
        "active_role": active_role or "",
    })
    print(json.dumps(output, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hooks/tool-guard.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import { trace } from './_trace.ts';
import { agentIsReadonly, agentToolsAllowlist, getActiveRole } from './_active_role.ts';
import { extractFilePath, extractToolName } from './_tool_payload.ts';

const EDIT_TOOLS = new Set(['Write', 'Edit', 'MultiEdit', 'NotebookEdit']);

const NAMING_SLOP_PATTERNS: Array<{ name: string; regex: RegExp }> = [
  { name: 'slop-final', regex: /-final(\.[^./]+)?$/ },
  { name: 'slop-final-vN', regex: /-final-v\d+(\.[^./]+)?$/ },
  { name: 'slop-backup', regex: /_backup\b/ },
  { name: 'slop-old', regex: /_old\b/ },
  { name: 'slop-copy', regex: /_copy\b/ },
  { name: 'slop-vN', regex: /_v\d+(\.[^./]+)?$/ },
  { name: 'slop-paren', regex: / \(\d+\)(\.[^./]+)?$/ },
  { name: 'slop-space-copy', regex: / copy(\.[^./]+)?$/ },
];

function matchNamingSlop(basename: string): string {
  for (const { name, regex } of NAMING_SLOP_PATTERNS) {
    if (regex.test(basename)) return name;
  }
  return '';
}

function readPayload(): any {
  try {
    const raw = fs.readFileSync(0, 'utf-8');
    if (!raw.trim()) {
      return {};
    }
    return JSON.parse(raw);
  } catch (err: any) {
    return { _invalid_json: true };
  }
}

function main(): number {
  const payload = readPayload();
  if (payload?._invalid_json) {
    console.log(JSON.stringify({
      status: 'pass',
      fail_open: true,
      permission: 'allow',
      user_message: 'Tool-guard input was not JSON; skipped.'
    }));
    return 0;
  }

  const toolName = extractToolName(payload);
  const filePath = extractFilePath(payload);
  const basename = filePath ? path.posix.basename(filePath.replace(/\\/g, '/')) : '';

  const activeRole = getActiveRole();
  const roleReadonly = activeRole ? agentIsReadonly(activeRole) : null;
  const roleTools = activeRole ? agentToolsAllowlist(activeRole) : null;

  let permission = 'allow';
  let message = 'Tool-guard saw no protected file mutation.';
  let status = 'pass';

  if (EDIT_TOOLS.has(toolName) && basename === 'workflow-state.json') {
    permission = 'ask';
    message = (
      'Tool-guard observed a direct edit to a workflow-state.json document. ' +
      'Prefer the cursor-state-bridge tools or `node --experimental-strip-types scripts/workflow-state.ts` ' +
      '(the installed `.cursor/state/workflow-state.ts` path is a compatibility shim) so phase, ' +
      'acceptance criteria, and history advance through one bounded path.'
    );
    status = 'ask';
  } else if (activeRole && roleReadonly === true && EDIT_TOOLS.has(toolName)) {
    permission = 'ask';
    message = (
      `Tool-guard: active subagent role \`${activeRole}\` declares \`readonly: true\` ` +
      `in agents/${activeRole}.md. Edit-class tools require user confirmation. ` +
      'If this edit is intended, approve to proceed; otherwise route the change to a non-readonly role.'
    );
    status = 'ask';
  } else if (activeRole && Array.isArray(roleTools) && toolName && !roleTools.includes(toolName)) {
    permission = 'ask';
    message = (
      `Tool-guard: active subagent role \`${activeRole}\` does not list \`${toolName}\` in its ` +
      `\`tools:\` allowlist (agents/${activeRole}.md). Approve to proceed or extend the allowlist.`
    );
    status = 'ask';
  } else if (EDIT_TOOLS.has(toolName) && basename && matchNamingSlop(basename)) {
    const slopName = matchNamingSlop(basename);
    permission = 'ask';
    status = 'ask';
    message = `Tool-guard observed an edit to \`${basename}\` whose name matches the \`${slopName}\` slop pattern. Prefer overwriting the canonical file instead of creating a slop variant. Approve if this is intentional (one-off backup, migration step).`;
  }

  const output = {
    status,
    fail_open: permission !== 'deny',
    permission,
    user_message: message,
    tool_name: toolName,
    file_path: filePath,
  };

  trace({
    hook: 'tool-guard',
    event: 'preToolUse',
    status,
    permission,
    tool_name: toolName,
    file_basename: basename,
    active_role: activeRole || '',
    naming_slop: matchNamingSlop(basename),
  });

  console.log(JSON.stringify(output));
  return 0;
}

process.exit(main());
